snowdrop-block/src/Snowdrop/Block/Application.hs

```diff
@@ -87,25 +117,32 @@ applyBlockImpl checkBIV osParams (BlkStateConfiguration {..}) rawBlk blk@Block{.
 -- 3. for each block in the fork: payload is applied, blund is stored and tip updated.
 tryApplyFork
     -- TODO `undo` is not Monoid, even for ChangeSet
-    :: ( HasBlock header payload bdata
-       , HasGetter bdata rawPayload
-       , Eq blockRef
-       , HasExceptions e [ForkVerificationException blockRef, BlockApplicationException blockRef]
-       , MonadError e m
-       )
-    => BlkStateConfiguration header payload rawPayload undo blockRef m
+    :: forall header payload rawBlock rawPayload blockRef undo e m
+    . ( HasGetter rawBlock rawPayload
+      , HasBlock header payload  (rawPayload, (Block header payload))
+      , HasGetter (rawPayload, Block header payload) rawPayload
+      , Eq blockRef
+      , HasExceptions e [ForkVerificationException blockRef, BlockApplicationException blockRef]
+      , MonadError e m
+      )
+    => BlkStateConfiguration header payload rawBlock rawPayload undo blockRef m
     -> OSParams
-    -> OldestFirst NonEmpty bdata
+    -> OldestFirst NonEmpty rawBlock
     -> m Bool
-tryApplyFork bcs@(BlkStateConfiguration {..}) osParams fork = verifyFork bcs osParams fork >>= \case
+tryApplyFork bcs@(BlkStateConfiguration {..}) osParams rawBlocks = do
+  fork <- traverse toFork (unOldestFirst rawBlocks)
```

Comment:
no, it's wrong: you try to expand fork-blocks over a current state, not over an old state. 

also, there shouldn't be operation like "expand sequence of block using `bscExpand` and then apply them sequentially" at all in code because expanding a block depends on application of the previous block, so expanding of block should be done over the state which is formed by application of previous blocks.